server/src/main/java/org/apache/druid/server/coordinator/duty/MarkDanglingTombstonesAsUnused.java

```diff
@@ -0,0 +1,147 @@
+ * can have at most two dangling segments.
+ */
+public class MarkDanglingTombstonesAsUnused implements CoordinatorDuty
+{
+  private static final Logger log = new Logger(MarkDanglingTombstonesAsUnused.class);
+
+  private final SegmentDeleteHandler deleteHandler;
+
+  public MarkDanglingTombstonesAsUnused(final SegmentDeleteHandler deleteHandler)
+  {
+    this.deleteHandler = deleteHandler;
+  }
+
+  @Override
+  public DruidCoordinatorRuntimeParams run(final DruidCoordinatorRuntimeParams params)
+  {
+    DataSourcesSnapshot dataSourcesSnapshot = params.getDataSourcesSnapshot();
+
+    final Map<String, Set<SegmentId>> datasourceToDanglingTombstones = determineDanglingTombstones(dataSourcesSnapshot);
+
+    if (datasourceToDanglingTombstones.size() == 0) {
+      log.debug("No dangling tombstones found.");
+      return params;
+    }
+
+    log.debug("Found [%d] datasource dangling tombstones  [%s]",
+              datasourceToDanglingTombstones.size(), datasourceToDanglingTombstones
+    );
+
+    final CoordinatorRunStats stats = params.getCoordinatorStats();
+    datasourceToDanglingTombstones.forEach((datasource, unusedSegments) -> {
+      RowKey datasourceKey = RowKey.of(Dimension.DATASOURCE, datasource);
+      stats.add(Stats.Segments.DANGLING_TOMBSTONE, datasourceKey, unusedSegments.size());
+      int unusedCount = deleteHandler.markSegmentsAsUnused(unusedSegments);
+      log.info(
+          "Successfully marked [%d] dangling tombstones as unused for datasource[%s].",
+          unusedCount,
+          datasource
+      );
+    });
+
+    return params;
+  }
+
+  /**
+   * Computes the set of dangling tombstones using the datasources snapshot.
+   *
+   * <li> Determine the set of used and non-overshadowed segments from the used segments' timeline. </li>
+   * <li> For each such segment that is dangling, look at the set of used segments in the timeline to see if the intervals overlap.
+   * There can at most be two such dangling tombstones per datasource. </li>
+   * <li> If there is no overlap, add the dangling segment to the result set to be marked as unused. </li>
+   * </p>
+   *
+   * @param dataSourcesSnapshot the datasources snapshot for segments timeline
+   * @return the set of dangling tombstones grouped by datasource
+   */
+  private Map<String, Set<SegmentId>> determineDanglingTombstones(final DataSourcesSnapshot dataSourcesSnapshot)
+  {
+    final Map<String, Set<SegmentId>> datasourceToDanglingTombstones = new HashMap<>();
+
+    dataSourcesSnapshot.getDataSourcesMap().keySet().forEach((datasource) -> {
+      final SegmentTimeline usedSegmentsTimeline
+          = dataSourcesSnapshot.getUsedSegmentsTimelinesPerDataSource().get(datasource);
+
+
+      final Optional<Set<DataSegment>> usedNonOvershadowedSegments =
+          Optional.fromNullable(usedSegmentsTimeline)
+                  .transform(timeline -> timeline.findNonOvershadowedObjectsInInterval(
```

Comment:
Should we `findNonOvershadowed` again or should we just use the set of overshadowed segments captured in the `DataSourcesSnapshot`?

Reply:
Ah, yes, I considered just the overshadowed set of segments (please see c5fb0e9ccc823580f373e1a9da3091b6ddaba7bc), but realized that would not provide us a complete guarantee as there can be a partial overlap with a used segment that's not overshadowed, so we'd need to look at the larger set of `findNonOvershadowed` segments - added a UT for one such scenario `org.apache.druid.server.coordinator.duty.MarkDanglingSegmentsAsUnusedTest#testDanglingTombstonesWithPartiallyOverlappingUnderlyingSegment`.